hugr-passes/src/dataflow/partial_value.rs

```diff
@@ -0,0 +1,718 @@
+    /// whether `self` has changed.
+    ///
+    /// Fails (without mutation) with the conflicting tag if any common rows have different lengths.
+    pub fn try_join_mut(&mut self, other: Self) -> Result<bool, usize> {
+        for (k, v) in &other.0 {
+            if self.0.get(k).is_some_and(|row| row.len() != v.len()) {
+                return Err(*k);
+            }
+        }
+        let mut changed = false;
+
+        for (k, v) in other.0 {
+            if let Some(row) = self.0.get_mut(&k) {
+                for (lhs, rhs) in zip_eq(row.iter_mut(), v.into_iter()) {
+                    changed |= lhs.join_mut(rhs);
+                }
+            } else {
+                self.0.insert(k, v);
+                changed = true;
+            }
+        }
+        Ok(changed)
+    }
+
+    /// Mutates self according to lattice meet operation (towards `Bottom`). If successful,
+    /// returns whether `self` has changed.
+    ///
+    /// # Errors
+    /// Fails without mutation, either:
+    /// * `Some(tag)` if the two [PartialSum]s both had rows with that `tag` but of different lengths
+    /// * `None` if the two instances had no rows in common (i.e., the result is "Bottom")
+    pub fn try_meet_mut(&mut self, other: Self) -> Result<bool, Option<usize>> {
+        let mut changed = false;
+        let mut keys_to_remove = vec![];
+        for (k, v) in self.0.iter() {
+            match other.0.get(k) {
+                None => keys_to_remove.push(*k),
+                Some(o_v) => {
+                    if v.len() != o_v.len() {
+                        return Err(Some(*k));
+                    }
+                }
+            }
+        }
+        if keys_to_remove.len() == self.0.len() {
+            return Err(None);
+        }
+        for (k, v) in other.0 {
+            if let Some(row) = self.0.get_mut(&k) {
+                for (lhs, rhs) in zip_eq(row.iter_mut(), v.into_iter()) {
+                    changed |= lhs.meet_mut(rhs);
+                }
+            } else {
+                keys_to_remove.push(k);
+            }
+        }
+        for k in keys_to_remove {
+            self.0.remove(&k);
+            changed = true;
+        }
+        Ok(changed)
+    }
+
+    /// Whether this sum might have the specified tag
+    pub fn supports_tag(&self, tag: usize) -> bool {
+        self.0.contains_key(&tag)
+    }
+
+    /// Turns this instance into a [Sum] of some "concrete" value type `V2`,
+    /// *if* this PartialSum has exactly one possible tag.
+    ///
+    /// # Errors
+    ///
+    /// If this PartialSum had multiple possible tags; or if `typ` was not a [TypeEnum::Sum]
+    /// supporting the single possible tag with the correct number of elements and no row variables;
+    /// or if converting a child element failed via [PartialValue::try_into_value].
+    pub fn try_into_value<VE, SE, V2: TryFrom<V, Error = VE> + TryFrom<Sum<V2>, Error = SE>>(
```

Comment:
Yes, this is highly parametric. The first case will be the "easy" one, where `V2` is `hugr_core::ops::Value`, in which case `VE = Infallible` and `SE = ConstTypeError`. But I think it's important to keep this open for other analysis domains.

It might be more legible if written `<V2, VE, SE> ..... where V2: TryFrom<V, Error=VE> + TryFrom<Sum<V2>, Error=SE>>`, happy to do that if anyone shouts.

The bigger issue may be nomenclature - this relates to `ExtractValueError` so perhaps should be (`try_`)`extract_value`? And also to `Machine::try_read_wire_value` - so, `try_extract_wire_value`?? However I quite like the word "concrete" so could perhaps use that naming to tie the three together - "try_into_concrete", "try_read_wire_concrete", and, erm, "ConcretizationError" (hmmm, not so good....). Thoughts?

Reply:
> It might be more legible if written `<V2, VE, SE> ..... where V2: TryFrom<V, Error=VE> + TryFrom<Sum<V2>, Error=SE>>`, happy to do that if anyone shouts.

I think you should!

Reply:
Done - yes good to standardize the order of the args with `try_read_wire_value`.

Reply:
Any thoughts on naming?